src/zodbbrowser/references.py

```diff
@@ -0,0 +1,136 @@
+
+import sqlite3
+
+from ZODB.utils import u64
+from ZODB.serialize import referencesf
+from zope.interface import implements
+
+from zodbbrowser.interfaces import IReferencesDatabase
+
+
+def connect(callback):
+    """Decorator for the reference database to access the sqlite DB.
+    """
+
+    def wrapper(self, *args, **kwargs):
+        try:
+            connection = sqlite3.connect(self.db_name)
+        except:
+            raise ValueError('impossible to open references database')
+        try:
+            result = callback(self, connection, *args, **kwargs)
+        except:
+            connection.close()
+            raise
+        return result
+
+    return wrapper
+
+
+class ReferencesDatabase(object):
+    implements(IReferencesDatabase)
+
+    def __init__(self, db_name=':memory:'):
+        self.db_name = db_name
+
+    @connect
+    def analyzeRecords(self, connection, records):
+        cursor = connection.cursor()
+        for record in records:
+            current_oid = u64(record.oid)
+            referred_oids = set(map(u64, referencesf(record.data)))
+
+            for referred_oid in referred_oids or [-1]:
```

Comment:
It is used in order to mark that the object is present even if it does not have any references to an another. OIDs are always positive integers. Missing objects (poskey) will not be listed and won't appear in the relation database.